File: project_pyratio/algorithms.py
# algorithms.py
# Various smart functions for expanding factory plans in steps

from dataclasses import dataclass
from typing import Any

from databases import DataBase
from facilities import Facility
from factories import Factory, RecipeCrafter
from products import Product, ProductQuantity
from rates import Rate
from rational_utilities import try_read_rational
from recipes import Recipe
from user_questions import get_user_choice, get_user_text


###########################################
# Decisions that are Deferred to the User #
###########################################

# Utility return type for makeexpanddecision() function
@dataclass(frozen=True, slots=True)
class ExpandDecision:
    new_crafter: bool
    argument: Any


def make_expand_decision(
    factory: Factory,
    db: DataBase,
    p: Product
) -> ExpandDecision | None:
    existing_crafters = factory.find_crafters(p)
    possible_recipes = db.find_recipes(p)

    new_crafter: dict[str, bool] = {}
    args: dict[str, Any] = {}

    for crafter in existing_crafters:
        string = f"Expand Crafter: {crafter}"
        new_crafter[string] = False
        args[string] = crafter

    for recipe in possible_recipes:
        string = f"New Crafter: {recipe.name}"
        new_crafter[string] = True
        args[string] = recipe

    options = list(new_crafter.keys())
    chosen_string = get_user_choice(
        f"Choose how to produce {p}:",
        options,
        "<ignore this product>",
        sort_by=lambda x: x
    )

    if chosen_string is None:
        return None
    return ExpandDecision(new_crafter[chosen_string], args[chosen_string])


def make_facility_decision(db: DataBase, recipe: Recipe) -> Facility:
    possible_facilities = db.find_facilities(recipe)

    if len(possible_facilities) > 1:
        return get_user_choice(
            f"Select a facility to craft {recipe.name}:",
            possible_facilities,
            "<nevermind, ignore this product>",
            sortby=lambda x : x.speed
        )
    assert len(possible_facilities) == 1
    return possible_facilities.pop()


def make_goal_decision(db: DataBase) -> ProductQuantity[Rate] | None:
    item = get_user_choice(
        "Choose the product this factory should yield:",
        db.products,
        "<quit program>",
        sort_by=lambda x: x.name
    )
    if item is None:
        return None

    per_second = get_user_text(
        "Enter desired production rate (per second): ",
        try_read_rational
    )
    rate = Rate(per_second)

    # TODO: ask if we want the output product to be proliferated

    return ProductQuantity(rate, item)


def make_production_decision(
    factory: Factory,
    db: DataBase
) -> ProductQuantity[Rate] | None:
    possible_targets = find_production_gaps(factory, db)

    if len(possible_targets) == 0:
        return None
    return get_user_choice(
        "Choose next production target to satisfy:",
        possible_targets,
        "<ignore all and finish>",
        sort_by=lambda x: x.name
    )


####################################
# Algorithms for Factory Expansion #
####################################

def find_production_gaps(
    factory: Factory,
    db: DataBase
) -> list[ProductQuantity[Rate]]:
    possible_gaps = factory.negative_rates()
    verified_gaps: list[ProductQuantity[Rate]] = []

    for product, rate in possible_gaps.items():
        if len(db.find_recipes(product)) == 0:
            factory.set_ignored(product)
        else:
            verified_gaps.append(ProductQuantity(-rate, product))

    return verified_gaps


def expand_factory_once(
    factory: Factory,
    db: DataBase,
    objective: ProductQuantity[Rate]
) -> Factory:
    expand_decision = make_expand_decision(factory, db, objective.product)

    if expand_decision is None:
        factory.set_ignored(objective.product)
        return factory

    if expand_decision.new_crafter:  # Make a new crafter

        recipe: Recipe = expand_decision.argument
        facility = make_facility_decision(db, recipe)

        if facility is None:
            factory.set_ignored(objective.product)
            return factory

        new_crafter = RecipeCrafter.with_goal(recipe, facility, objective)
        factory.connect_crafter(new_crafter)

    else:  # Expand an existing crafter

        crafter: RecipeCrafter = expand_decision.argument
        factory.upgrade_crafter(crafter, objective)

    return factory


def generate_factory(db: DataBase) -> Factory | None:
    goal = make_goal_decision(db)
    if goal is None:
        return None
    factory = Factory(goal)

    while True:
        target = make_production_decision(factory, db)
        if target is None:
            factory.set_all_ignored()
            break
        expand_factory_once(factory, db, target)

    return factory

File: julia_code/algorithms.jl
# algorithms.jl
# Various smart functions for expanding factory plans in steps


###########################################
# Decisions that are Deferred to the User #
###########################################

# Utility return type for makeexpanddecision() function
struct ExpandDecision
	newcrafter::Bool
	argument::Any
end

function makeexpanddecision(factory::Factory, db::DataBase, p::Product)
	
	existingcrafters = findcrafters(factory, p)
	possiblerecipes = findrecipes(db, p)
	
	newcrafter = Dict{String, Bool}()
	args = Dict{String, Any}()
	
	for crafter in existingcrafters
		str = string("Expand Crafter: ", crafter)
		newcrafter[str] = false
		args[str] = crafter
	end
	
	for recipe in possiblerecipes
		str = string("New Crafter: ", name(recipe))
		newcrafter[str] = true
		args[str] = recipe
	end
	
	options = keys(newcrafter)
	chosenstring = getuserchoice("Choose how to produce $(p):", options,
		"<ignore this product>", sortby = identity)
	
	if isnothing(chosenstring)
		return nothing
	else
		return ExpandDecision(newcrafter[chosenstring], args[chosenstring])
	end
end

function makefacilitydecision(db::DataBase, recipe::Recipe)
	
	possiblefacilities = findfacilities(db, recipe)
	
	if length(possiblefacilities) > 1
		return getuserchoice("Select a facility to craft $(name(recipe)):",
			possiblefacilities, "<nevermind, ignore this product>",
			sortby = speed)
	else
		return only(possiblefacilities)
	end
end

function makegoaldecision(db::DataBase)
	
	item = getuserchoice("Choose the product this factory should yield:",
		db.products, "<quit program>", sortby = name)
	isnothing(item) && return nothing
	
	persecond = getusertext("Enter desired production rate (per second): ",
		tryreadrational)
	rate = Rate(persecond)
	
	# TODO: ask if we want the output product to be proliferated
	
	return ProductQuantity(rate, item)
end

function makeproductiondecision(factory::Factory, db::DataBase)
	
	possibletargets = findproductiongaps(factory, db)
	
	if isempty(possibletargets)
		return nothing
	else
		return getuserchoice("Choose next production target to satisfy:",
			possibletargets, "<ignore all and finish>", sortby = name)
	end
end

####################################
# Algorithms for Factory Expansion #
####################################

function findproductiongaps(factory::Factory, db::DataBase)
	
	possiblegaps = negativerates(factory)
	verifiedgaps = Vector{ProductQuantity{Rate}}()
	
	for (product, rate) in possiblegaps
		if isempty(findrecipes(db, product))
			setignored!(factory, product)
		else
			push!(verifiedgaps, ProductQuantity(-rate, product))
		end
	end
	
	verifiedgaps
end

function expandfactoryonce!(factory::Factory, db::DataBase,
	objective::ProductQuantity{Rate})
	
	expanddecision = makeexpanddecision(factory, db, product(objective))
	
	if isnothing(expanddecision)
		setignored!(factory, product(objective))
		return factory
	end
	
	if expanddecision.newcrafter  # Make a new crafter
		
		recipe = expanddecision.argument :: Recipe
		facility = makefacilitydecision(db, recipe)
		
		if isnothing(facility)
			setignored!(factory, product(objective))
			return factory
		end
		
		newcrafter = RecipeCrafter(recipe, facility, objective)
		connectcrafter!(factory, newcrafter)
	
	else # Expand an existing crafter
		
		crafter = expanddecision.argument :: RecipeCrafter
		upgradecrafter!(factory, crafter, objective)
	
	end
	
	return factory
end

function generatefactory(db::DataBase)
	
	goal = makegoaldecision(db)
	isnothing(goal) && return nothing
	factory = Factory(goal)
	
	while true
		target = makeproductiondecision(factory, db)
		if isnothing(target)
			setallignored!(factory)
			break
		else
			expandfactoryonce!(factory, db, target)
		end
	end
	
	factory
end

